Added health checks into our custom executors

elingg opened the pull request.

elingg commented:
@kensipe and @gabrielhartmann, please take a look when you have a chance. thx! I have tested failure scenarios on Mesos itself. I will need to test on DCOS before merging.

gabrielhartmann commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java:
```diff
@@ -29,6 +30,10 @@
   private Task zkfcNodeTask;
+  //Timed Health Check for node health monitoring
+  private TimedHealthCheck healthCheckNN;
+  private TimedHealthCheck healthCheckZN;
```
We don't need these private healthCheckNN members.  They can just be created when needed.  They aren't reference outside of the point at which they are launched.

elingg replied:
great point. yes

gabrielhartmann commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java:
```diff
@@ -66,12 +72,20 @@ public void launchTask(final ExecutorDriver driver, final TaskInfo taskInfo) {
           .build());
+      healthCheckNN = new TimedHealthCheck(driver, nameNodeTask);
+      timer.scheduleAtFixedRate(healthCheckNN,
+        hdfsFrameworkConfig.getHealthCheckWaitingPeriod(),
```
This can be timer.scheduleAtFixedRate(new TimedHealthcheck(driver, nameNodeTask), .... right?

elingg replied:
method chaining is a java style that we try to avoid as consistently as possible in this project. This is because it becomes harder to identify errors. It is a pretty common java style guideline.

gabrielhartmann commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/AbstractNodeExecutor.java:
```diff
@@ -289,6 +293,83 @@ private void sendTaskFailed(ExecutorDriver driver, Task task) {
+        }
+      }
+    }
+    if (!taskHealthy) {
```
Please add a newline before this if statement.

elingg replied:
fixed

gabrielhartmann commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/NodeExecutor.java:
```diff
@@ -20,6 +23,9 @@
   private final Log log = LogFactory.getLog(NodeExecutor.class);
   private Task task;
+  //Timed Health Check for node health monitoring
+  private TimedHealthCheck timedHealthCheck;
```
Same here, this private member isn't needed.

elingg replied:
true

gabrielhartmann commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/NodeExecutor.java:
```diff
@@ -51,6 +57,11 @@ public void launchTask(final ExecutorDriver driver, final TaskInfo taskInfo) {
         .setData(taskInfo.getData()).build());
+    timer = new Timer(true);
+    timedHealthCheck = new TimedHealthCheck(driver, task);
+    timer.scheduleAtFixedRate(timedHealthCheck,
```
Again, TimedHealthCheck can be constructed on demand and doesn't need to be stored in a private member variable.

elingg replied:
fixed

gabrielhartmann commented:
If I'm reading this correctly, if a Task's port is unexpectedly available after 60, we assume that the Task is unhealthy and should be killed correct?  Then the Scheduler will be notified and we'll restart it.

Why did you choose the port availability for the health check?  It might be better to go the HTTP route which will soon be available anyway.  You could do your own checks against the HTTP interface and then when Mesos supports HTTP healthcheck's very little work will remain.

I'd also prefer using HTTP to check health because it's unambiguous.  Return appropriate error codes and we are getting real signals from the Tasks as opposed to making assumptions based on port availability.

Other than that, I had only very minor comments about the code itself.

elingg commented:
Thanks @gabrielhartmann, unfortunately the http and mesos task health checks are implemented for the default command executor, which is why I had to implement this for our custom executor. I hope this will be updated in Mesos, but it is not done yet and I cannot wait on it.

gabrielhartmann commented:
@elingg:  I know the http health checks aren't available yet, and I definitely don't think we should wait.  I'm just saying having the Tasks provide HTTP endpoints that we can query for health checks as opposed to making assumptions based on port availability is preferable.  It also has the added benefit, that if you expose HTTP endpoints on the Tasks now it will be trivial to consume the Mesos HTTP health checks when they arrive.

elingg commented:
Discussed with @gabrielhartmann in person about determined that ZKFC does not have a webui and that health checks are implemented for the mesos command executor but not custom executors. 

I will revise the IOException to check and see if the port is in use. I will also evaluate the initial timeout period to see if it's long enough for a typical bootstrap scenario. I will test the health check on the second NN that needs to be bootstrapped (the easiest way is killing the ZKFC) to make sure it fails the health check and then comes back up. I also will test on DCOS.

kensipe commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/AbstractNodeExecutor.java:
```diff
@@ -28,10 +28,14 @@
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.TimerTask;
```
I already know the push back on this :)  however.... I would to see smaller classes.  This seems like functionality that would be in a NodeHealthChecker class.

gabrielhartmann replied:
+1

elingg replied:
Sure, I also like smaller classes. Let me pull this out. Don't foresee any issues at least.

kensipe commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java:
```diff
@@ -29,6 +30,8 @@
   // chained.
   private Task zkfcNodeTask;
+  //Timed Health Check for node health monitoring
+  private Timer healthCheckTimer;
```
If we are going to have health checks in the abstract base... shouldn't the healthCheckTimer be protected in the base?  the code is the same in both executors.  The only difference is the creation of the the scheduled health check.

I would even be inclined to add a function in the base for `scheduleHealthCheck(...)`  

This Timer and the initialization are the same in both executors.

elingg replied:
yes, true, if we are going to keep it in the base. I will be pulling it out though.

kensipe commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/AbstractNodeExecutor.java:
```diff
@@ -289,6 +293,84 @@ private void sendTaskFailed(ExecutorDriver driver, Task task) {
+    } else if (taskIdStr.contains(HDFSConstants.NAME_NODE_ID)) {
+      healthCheckPort = HDFSConstants.NAME_NODE_PORT;
+    } else {
+      log.error("Task unknown: " + taskIdStr);
```
we are starting to see this pattern in a number of places... I think after you commit I will refactor a number of these to a hashmap of config objects.  It will look cleaner IMO.

elingg replied:
That sounds like an awesome refactoring if we can clean this up further.

kensipe commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/AbstractNodeExecutor.java:
```diff
@@ -289,6 +293,84 @@ private void sendTaskFailed(ExecutorDriver driver, Task task) {
+      } finally {
+        if (socket != null) {
+          try {
+            socket.close();
```
This really needs to be a `IOUtils.close()`  It would remove many lines of code.
(this is the apache io which is already in the project.

elingg replied:
Great point, done.

kensipe commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/AbstractNodeExecutor.java:
```diff
@@ -289,6 +293,84 @@ private void sendTaskFailed(ExecutorDriver driver, Task task) {
+    
+    if (!taskHealthy) {
+      log.fatal("Node health check failed for task: " + taskIdStr);
+      killTask(driver, task.getTaskInfo().getTaskId());
```
I'm wondering... if the namenode is killed... don't we want the zk node to be killed as well?

elingg replied:
Yes, we want to. And this is why we actually terminate the executor which will kill both tasks. I have a note here that with better process supervision, we should not have to kill the full executor, but I think there's work to be done there including monitoring the ZK task in our persistent and live state and I'd rather play it safe in the meantime.

elingg commented:
I ended up splitting some health check logic out, but kept the calls to that logic in the AbstractExecutor. I have addressed your comments.

I still need to do more extensive testing of failure scenarios (NN1 health check fails, NN2 health check fails, DCOS test), but other than that the review comments should be addressed.

@kensipe, @gabrielhartmann, if you could take a final look and determine if your comments have been addressed, that would be great! I will ultimately squash the commits and merge once ready.

gabrielhartmann commented:
LGTM

elingg commented:
Tested failure scenarios more extensively and tested on Mesos and DCOS. Also, removed a check after seeing test results. Ready for merge. Please let me know if there are any objections, otherwise I will go ahead and merge on Monday @kensipe and @gabrielhartmann.

elingg commented:
Tested failure scenarios more extensively and tested on Mesos and DCOS. Also, removed a check and fixed shutdown of the executor after seeing test results. Ready for merge. Please take a second look if you can and let me know if there are any objections, otherwise I will go ahead and merge on Monday @kensipe and @gabrielhartmann.

elingg commented:
Due to a number of changes in master recently, had a few rebase issues. Will simply merge with the full commit history, which will include the review comments for some level of clarity
